source/extensions/http/stateful_session/cookie/cookie.cc

```diff
@@ -11,18 +12,33 @@ namespace Cookie {
 void CookieBasedSessionStateFactory::SessionStateImpl::onUpdate(
     const Upstream::HostDescription& host, Envoy::Http::ResponseHeaderMap& headers) {
   absl::string_view host_address = host.address()->asStringView();
+  std::string encoded_address;
   if (!upstream_address_.has_value() || host_address != upstream_address_.value()) {
-    const std::string encoded_address =
-        Envoy::Base64::encode(host_address.data(), host_address.length());
+    if (Runtime::runtimeFeatureEnabled(
+            "envoy.reloadable_features.stateful_session_encode_ttl_in_cookie") &&
+        !use_old_style_encoding_) {
```

Comment:
The purpose of this runtime guard is for users which interpret cookie in some way and want Envoy to keep sending cookie in the "old" format. If there is no runtime guard here, there is no way to force Envoy to send cookies in "old" format.

Reply:
I think there is a gap about the back compatibility. When I say we should keep the back compatibility, I mean that both the new format cookie or old format cookie could be parsed by the envoy.
We never expect that the cookie be parsed by others. Just like the cookie generated by the consistent hash lb. It should be meaningless for others except the envoy.

And I wonder in this case, when we can deprecate the old format cookie **generation** completely?

Even this runtime flag is removed in the future, if the clients still send requests with invalid old format cookie, the envoy will set a valid old format cookie to the clients again. Then these clients will use old format cookie until to forever.

Reply:
> I think there is a gap about the back compatibility. When I say we should keep the back compatibility, I mean that both the new format cookie or old format cookie could be parsed by the envoy.
We never expect that the cookie be parsed by others. Just like the cookie generated by the consistent hash lb. It should be meaningless for others except the envoy.
(But if some one expect it, it ok for me 😉)

Actually I think we are on the same page. The proper interaction with a client assumes that a client sends the first request without a cookie. Envoy selects upstream host and generates cookie which is sent back to the client. Since that moment the client attaches the received cookie to subsequent requests and is always forwarded to the same upstream host. In this "ideal" scenario, backwards compatibility is not required. Envoy will always generate cookie in new format. Clients, as you said, should not care about the cookie internal format. However, as noted in https://github.com/envoyproxy/envoy/issues/26347#issuecomment-1503506078, sometimes people may want to influence where the first request is routed. I think it is incorrect approach to using cookie, but who knows what happens in the field. For that reason, "old" format is accepted.

> And I wonder in this case, when we can deprecate the old format cookie generation completely?

I added a log to indicate that "old" non-proto format will be deprecated in the future. It should happen 12 months after PR is merged. Again, "normal" users should never see that log. It will pop-up only when someone manually generates the cookie and attaches it to requests.

Reply:
> I added a log to indicate that "old" non-proto format will be deprecated in the future. It should happen 12 months after PR is merged. Again, "normal" users should never see that log. It will pop-up only when someone manually generates the cookie and attaches it to requests.

I think the normal users that have used the old format cookie will also see the log always and it would brings lots of performance overhead?